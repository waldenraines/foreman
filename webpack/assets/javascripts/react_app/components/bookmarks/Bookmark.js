import React from 'react';
import URI from 'urijs';
import { MenuItem } from 'react-bootstrap';

function followLink() {
  const uri = new URI(window.location.href);

  uri.setSearch('search', this.trim());
  window.Turbolinks.visit(uri.toString());
  return false;
}

const Bookmark = ({ text, query }) => <MenuItem onClick={followLink.bind(query)}>{text}</MenuItem>;

export default Bookmark;
