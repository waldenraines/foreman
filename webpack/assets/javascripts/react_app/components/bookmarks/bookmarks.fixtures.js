/* eslint-disable camelcase */
import Immutable from 'seamless-immutable';

export const initialState = Immutable({
  bookmarks: {
    showModal: false,
  },
});
export const bookmarks = [
  {
    name: '1111',
    controller: 'hosts',
    query: 'abc',
    public: true,
    id: 52,
    owner_id: 1,
    owner_type: 'User',
  },
  {
    name: '1122',
    controller: 'hosts',
    query: 'abc',
    public: true,
    id: 54,
    owner_id: 1,
    owner_type: 'User',
  },
];

export const afterSuccess = Immutable({
  bookmarks: {
    ...initialState,
    hosts: bookmarks,
  },
});
