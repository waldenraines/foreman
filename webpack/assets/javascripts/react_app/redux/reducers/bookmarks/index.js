import Immutable from 'seamless-immutable';
import {
  BOOKMARKS_REQUEST,
  BOOKMARKS_SUCCESS,
  BOOKMARKS_FAILURE,
  BOOKMARKS_MODAL_OPENED,
  BOOKMARKS_MODAL_CLOSED,
  BOOKMARK_FORM_SUBMITTED,
} from '../../consts';

const initialState = Immutable({
  showModal: false,
});

const sortByName = (a, b) => {
  if (a.name < b.name) {
    return -1;
  }
  if (a.name > b.name) {
    return 1;
  }
  // names must be equal
  return 0;
};

export default (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case BOOKMARKS_REQUEST:
      return state.set(payload.controller, []);
    case BOOKMARKS_SUCCESS:
      return state.set(payload.controller, payload.results);
    case BOOKMARKS_MODAL_OPENED:
      return state.set('currentQuery', payload.query).set('showModal', true);
    case BOOKMARK_FORM_SUBMITTED:
      return state
        .set(
          payload.body.controller,
          [...state[payload.body.controller], payload.body].sort(sortByName),
        )
        .set('showModal', false);
    case BOOKMARKS_MODAL_CLOSED:
      return state.set('showModal', false);
    case BOOKMARKS_FAILURE:
    default:
      return state;
  }
};
