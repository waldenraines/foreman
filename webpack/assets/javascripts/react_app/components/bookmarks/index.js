import React from 'react';
import { connect } from 'react-redux';
import { Dropdown, MenuItem } from 'react-bootstrap';
import SearchModal from './SearchModal';
import Bookmark from './Bookmark';
import * as BookmarkActions from '../../redux/actions/bookmarks';
import DocumentationUrl from '../common/DocumentationLink';

class BookmarkContainer extends React.Component {
  componentDidMount() {
    const { url, controller, getBookmarks } = this.props;

    getBookmarks(url, controller);
  }

  onClick() {
    if (this.props.showModal) {
      this.props.modalClosed();
    } else {
      this.props.modalOpened();
    }
  }

  render() {
    const {
      controller, url, showModal, modalClosed, canCreate,
    } = this.props;

    return showModal ? (
      <SearchModal show={showModal} controller={controller} url={url} onHide={modalClosed} />
    ) : (
      <Dropdown pullRight id={controller}>
        <Dropdown.Toggle />
        <Dropdown.Menu>
          <MenuItem header>{__('Saved Bookmarks')}</MenuItem>
          {this.props.bookmarks.map(({ name, query }) => (
            <Bookmark key={name} text={name} query={query} />
          ))}
          <MenuItem divider={true} />
          {canCreate && (
            <MenuItem key="newBookmark" id="newBookmark" onClick={this.onClick.bind(this)}>
              {__('Bookmark this search')}
            </MenuItem>
          )}
          <DocumentationUrl id="bookmarkDocumentation" href={this.props.documentationUrl} />
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

const mapStateToProps = ({ bookmarks }, { data: { controller } }) => ({
  bookmarks: bookmarks[controller] || [],
  showModal: bookmarks.showModal,
});

// I'm flatting the props that come from data attribute, this is done to avoid
// having special handling for this.props.data vs this.props in the codebase.
const mergeProps = (stateProps, dispatchProps, { data }) =>
  Object.assign(stateProps, data, dispatchProps);

export default connect(mapStateToProps, BookmarkActions, mergeProps)(BookmarkContainer);
