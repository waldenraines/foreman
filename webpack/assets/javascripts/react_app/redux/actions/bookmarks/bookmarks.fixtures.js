import immutable from 'seamless-immutable';
import * as types from '../../consts';
import { bookmarks } from '../../reducers/bookmarks/bookmarks.fixtures';

export const initialState = immutable({
  bookmarks: {
    hosts: [],
    showModal: false,
  },
});

export const requestData = { url: '/api/bookmarks', controller: 'hosts' };

const requestAction = {
  type: types.BOOKMARKS_REQUEST,
  payload: { controller: 'hosts' },
};

export const onFailureActions = [
  requestAction,
  {
    payload: { error: {}, item: { controller: 'hosts' } },
    type: types.BOOKMARKS_FAILURE,
  },
];

export const onSuccessActions = [
  requestAction,
  {
    payload: {
      results: bookmarks,
      controller: 'hosts',
    },
    type: types.BOOKMARKS_SUCCESS,
  },
];
