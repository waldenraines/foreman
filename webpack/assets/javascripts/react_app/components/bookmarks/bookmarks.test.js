import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import BookmarksContainer from './';
import { initialState, afterSuccess } from './bookmarks.fixtures';

const mockStore = configureMockStore([thunk]);

function setup() {
  const props = {
    data: {
      controller: 'hosts',
      url: '/api/bookmarks',
      canCreate: true,
    },
  };

  const wrapper = mount(<Provider store={mockStore(initialState)}>
      <BookmarksContainer {...props} />
    </Provider>);

  return {
    props,
    wrapper,
  };
}

describe('bookmarks', () => {
  it('empty state', () => {
    const { wrapper } = setup();

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should include existing bookmarks for the current controller', () => {
    const wrapper = mount(<Provider store={mockStore(afterSuccess)}>
        <BookmarksContainer {...setup().props} />
      </Provider>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should not allow creating a new bookmark for users who dont have permission', () => {
    const { props } = setup();

    props.canCreate = false;

    const wrapper = mount(<Provider store={mockStore(afterSuccess)}>
        <BookmarksContainer {...props} />
      </Provider>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should hide the modal form intiailly', () => {
    const { wrapper } = setup();

    expect(wrapper.find('BookmarkContainer').props().showModal).toBe(false);
  });
  xit('should open the modal form for new bookmark', () => {
    const { wrapper } = setup();

    expect(wrapper.find('BookmarkContainer').props().showModal).toBe(false);
    wrapper.find('a #newBookmark').simulate('click');
    expect(wrapper.find('BookmarkContainer').props().showModal).toBe(true);
    // TODO: look at alternative solution at https://github.com/airbnb/enzyme/issues/252#issuecomment-266125422
  });
  xit('full flow', () => {
    const { wrapper } = setup();

    expect(wrapper.find('Bookmark').length).toEqual(0);
    wrapper.find('a #newBookmark').simulate('click');

    const formWrapper = wrapper.find('SearchModal');

    formWrapper.find('input [name="name"]').simulate('change', { target: { value: 'Joe.D' } });
    formWrapper.find('textarea [name="query"]').simulate('change', { target: { value: 'search' } });
    formWrapper.find('input [name="publik"]').simulate('change', { target: { value: false } });

    formWrapper.find('form').simulate('submit');
    expect(wrapper.find('Bookmark').length).toEqual(1);
  });
});
