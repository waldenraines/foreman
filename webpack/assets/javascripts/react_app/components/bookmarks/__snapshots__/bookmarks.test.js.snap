// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookmarks empty state 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(BookmarkContainer)
    data={
      Object {
        "canCreate": true,
        "controller": "hosts",
        "url": "/api/bookmarks",
      }
    }
  >
    <BookmarkContainer
      bookmarks={Array []}
      canCreate={true}
      controller="hosts"
      getBookmarks={[Function]}
      modalClosed={[Function]}
      modalOpened={[Function]}
      showModal={false}
      url="/api/bookmarks"
    >
      <Uncontrolled(Dropdown)
        id="hosts"
        pullRight={true}
      >
        <Dropdown
          bsClass="dropdown"
          componentClass={[Function]}
          id="hosts"
          onToggle={[Function]}
          pullRight={true}
        >
          <ButtonGroup
            block={false}
            bsClass="btn-group"
            className="dropdown"
            justified={false}
            vertical={false}
          >
            <div
              className="dropdown btn-group"
            >
              <DropdownToggle
                bsClass="dropdown-toggle"
                bsRole="toggle"
                id="hosts"
                key=".0"
                onClick={[Function]}
                onKeyDown={[Function]}
                open={false}
                useAnchor={false}
              >
                <Button
                  active={false}
                  aria-expanded={false}
                  aria-haspopup={true}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="dropdown-toggle"
                  disabled={false}
                  id="hosts"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-default"
                    disabled={false}
                    id="hosts"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    type="button"
                  >
                     
                    <span
                      className="caret"
                    />
                  </button>
                </Button>
              </DropdownToggle>
              <DropdownMenu
                bsClass="dropdown-menu"
                bsRole="menu"
                key=".1"
                labelledBy="hosts"
                onClose={[Function]}
                onSelect={[Function]}
                pullRight={true}
              >
                <RootCloseWrapper
                  disabled={true}
                  event="click"
                  onRootClose={[Function]}
                >
                  <ul
                    aria-labelledby="hosts"
                    className="dropdown-menu dropdown-menu-right"
                    role="menu"
                  >
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={false}
                      header={true}
                      key=".0"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className="dropdown-header"
                        onKeyDown={[Function]}
                        role="heading"
                      >
                        Saved Bookmarks
                      </li>
                    </MenuItem>
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={true}
                      header={false}
                      key=".2"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className="divider"
                        onKeyDown={[Function]}
                        role="separator"
                      />
                    </MenuItem>
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={false}
                      header={false}
                      id="newBookmark"
                      key=".$newBookmark"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className=""
                        role="presentation"
                      >
                        <SafeAnchor
                          componentClass="a"
                          id="newBookmark"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          role="menuitem"
                          tabIndex="-1"
                        >
                          <a
                            href="#"
                            id="newBookmark"
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            Bookmark this search
                          </a>
                        </SafeAnchor>
                      </li>
                    </MenuItem>
                    <Component
                      id="bookmarkDocumentation"
                      key=".4"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        id="bookmarkDocumentation"
                        key="documentationUrl"
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            id="bookmarkDocumentation"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              id="bookmarkDocumentation"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              <Component
                                className="icon-black"
                                type="question-sign"
                              >
                                <span
                                  className="glyphicon glyphicon-question-sign icon-black"
                                />
                              </Component>
                               Documentation
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Component>
                  </ul>
                </RootCloseWrapper>
              </DropdownMenu>
            </div>
          </ButtonGroup>
        </Dropdown>
      </Uncontrolled(Dropdown)>
    </BookmarkContainer>
  </Connect(BookmarkContainer)>
</Provider>
`;

exports[`bookmarks should include existing bookmarks for the current controller 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(BookmarkContainer)
    data={
      Object {
        "canCreate": true,
        "controller": "hosts",
        "url": "/api/bookmarks",
      }
    }
  >
    <BookmarkContainer
      bookmarks={
        Array [
          Object {
            "controller": "hosts",
            "id": 52,
            "name": "1111",
            "owner_id": 1,
            "owner_type": "User",
            "public": true,
            "query": "abc",
          },
          Object {
            "controller": "hosts",
            "id": 54,
            "name": "1122",
            "owner_id": 1,
            "owner_type": "User",
            "public": true,
            "query": "abc",
          },
        ]
      }
      canCreate={true}
      controller="hosts"
      getBookmarks={[Function]}
      modalClosed={[Function]}
      modalOpened={[Function]}
      url="/api/bookmarks"
    >
      <Uncontrolled(Dropdown)
        id="hosts"
        pullRight={true}
      >
        <Dropdown
          bsClass="dropdown"
          componentClass={[Function]}
          id="hosts"
          onToggle={[Function]}
          pullRight={true}
        >
          <ButtonGroup
            block={false}
            bsClass="btn-group"
            className="dropdown"
            justified={false}
            vertical={false}
          >
            <div
              className="dropdown btn-group"
            >
              <DropdownToggle
                bsClass="dropdown-toggle"
                bsRole="toggle"
                id="hosts"
                key=".0"
                onClick={[Function]}
                onKeyDown={[Function]}
                open={false}
                useAnchor={false}
              >
                <Button
                  active={false}
                  aria-expanded={false}
                  aria-haspopup={true}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="dropdown-toggle"
                  disabled={false}
                  id="hosts"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-default"
                    disabled={false}
                    id="hosts"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    type="button"
                  >
                     
                    <span
                      className="caret"
                    />
                  </button>
                </Button>
              </DropdownToggle>
              <DropdownMenu
                bsClass="dropdown-menu"
                bsRole="menu"
                key=".1"
                labelledBy="hosts"
                onClose={[Function]}
                onSelect={[Function]}
                pullRight={true}
              >
                <RootCloseWrapper
                  disabled={true}
                  event="click"
                  onRootClose={[Function]}
                >
                  <ul
                    aria-labelledby="hosts"
                    className="dropdown-menu dropdown-menu-right"
                    role="menu"
                  >
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={false}
                      header={true}
                      key=".0"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className="dropdown-header"
                        onKeyDown={[Function]}
                        role="heading"
                      >
                        Saved Bookmarks
                      </li>
                    </MenuItem>
                    <Bookmark
                      key=".1:$1111"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                      query="abc"
                      text="1111"
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              1111
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Bookmark>
                    <Bookmark
                      key=".1:$1122"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                      query="abc"
                      text="1122"
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              1122
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Bookmark>
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={true}
                      header={false}
                      key=".2"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className="divider"
                        onKeyDown={[Function]}
                        role="separator"
                      />
                    </MenuItem>
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={false}
                      header={false}
                      id="newBookmark"
                      key=".$newBookmark"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className=""
                        role="presentation"
                      >
                        <SafeAnchor
                          componentClass="a"
                          id="newBookmark"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          role="menuitem"
                          tabIndex="-1"
                        >
                          <a
                            href="#"
                            id="newBookmark"
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            Bookmark this search
                          </a>
                        </SafeAnchor>
                      </li>
                    </MenuItem>
                    <Component
                      id="bookmarkDocumentation"
                      key=".4"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        id="bookmarkDocumentation"
                        key="documentationUrl"
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            id="bookmarkDocumentation"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              id="bookmarkDocumentation"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              <Component
                                className="icon-black"
                                type="question-sign"
                              >
                                <span
                                  className="glyphicon glyphicon-question-sign icon-black"
                                />
                              </Component>
                               Documentation
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Component>
                  </ul>
                </RootCloseWrapper>
              </DropdownMenu>
            </div>
          </ButtonGroup>
        </Dropdown>
      </Uncontrolled(Dropdown)>
    </BookmarkContainer>
  </Connect(BookmarkContainer)>
</Provider>
`;

exports[`bookmarks should not allow creating a new bookmark for users who dont have permission 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(BookmarkContainer)
    canCreate={false}
    data={
      Object {
        "canCreate": true,
        "controller": "hosts",
        "url": "/api/bookmarks",
      }
    }
  >
    <BookmarkContainer
      bookmarks={
        Array [
          Object {
            "controller": "hosts",
            "id": 52,
            "name": "1111",
            "owner_id": 1,
            "owner_type": "User",
            "public": true,
            "query": "abc",
          },
          Object {
            "controller": "hosts",
            "id": 54,
            "name": "1122",
            "owner_id": 1,
            "owner_type": "User",
            "public": true,
            "query": "abc",
          },
        ]
      }
      canCreate={true}
      controller="hosts"
      getBookmarks={[Function]}
      modalClosed={[Function]}
      modalOpened={[Function]}
      url="/api/bookmarks"
    >
      <Uncontrolled(Dropdown)
        id="hosts"
        pullRight={true}
      >
        <Dropdown
          bsClass="dropdown"
          componentClass={[Function]}
          id="hosts"
          onToggle={[Function]}
          pullRight={true}
        >
          <ButtonGroup
            block={false}
            bsClass="btn-group"
            className="dropdown"
            justified={false}
            vertical={false}
          >
            <div
              className="dropdown btn-group"
            >
              <DropdownToggle
                bsClass="dropdown-toggle"
                bsRole="toggle"
                id="hosts"
                key=".0"
                onClick={[Function]}
                onKeyDown={[Function]}
                open={false}
                useAnchor={false}
              >
                <Button
                  active={false}
                  aria-expanded={false}
                  aria-haspopup={true}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="dropdown-toggle"
                  disabled={false}
                  id="hosts"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-default"
                    disabled={false}
                    id="hosts"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    type="button"
                  >
                     
                    <span
                      className="caret"
                    />
                  </button>
                </Button>
              </DropdownToggle>
              <DropdownMenu
                bsClass="dropdown-menu"
                bsRole="menu"
                key=".1"
                labelledBy="hosts"
                onClose={[Function]}
                onSelect={[Function]}
                pullRight={true}
              >
                <RootCloseWrapper
                  disabled={true}
                  event="click"
                  onRootClose={[Function]}
                >
                  <ul
                    aria-labelledby="hosts"
                    className="dropdown-menu dropdown-menu-right"
                    role="menu"
                  >
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={false}
                      header={true}
                      key=".0"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className="dropdown-header"
                        onKeyDown={[Function]}
                        role="heading"
                      >
                        Saved Bookmarks
                      </li>
                    </MenuItem>
                    <Bookmark
                      key=".1:$1111"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                      query="abc"
                      text="1111"
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              1111
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Bookmark>
                    <Bookmark
                      key=".1:$1122"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                      query="abc"
                      text="1122"
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              1122
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Bookmark>
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={true}
                      header={false}
                      key=".2"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className="divider"
                        onKeyDown={[Function]}
                        role="separator"
                      />
                    </MenuItem>
                    <MenuItem
                      bsClass="dropdown"
                      disabled={false}
                      divider={false}
                      header={false}
                      id="newBookmark"
                      key=".$newBookmark"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <li
                        className=""
                        role="presentation"
                      >
                        <SafeAnchor
                          componentClass="a"
                          id="newBookmark"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          role="menuitem"
                          tabIndex="-1"
                        >
                          <a
                            href="#"
                            id="newBookmark"
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            Bookmark this search
                          </a>
                        </SafeAnchor>
                      </li>
                    </MenuItem>
                    <Component
                      id="bookmarkDocumentation"
                      key=".4"
                      onKeyDown={[Function]}
                      onSelect={[Function]}
                    >
                      <MenuItem
                        bsClass="dropdown"
                        disabled={false}
                        divider={false}
                        header={false}
                        id="bookmarkDocumentation"
                        key="documentationUrl"
                        onClick={[Function]}
                      >
                        <li
                          className=""
                          role="presentation"
                        >
                          <SafeAnchor
                            componentClass="a"
                            id="bookmarkDocumentation"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex="-1"
                          >
                            <a
                              href="#"
                              id="bookmarkDocumentation"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              <Component
                                className="icon-black"
                                type="question-sign"
                              >
                                <span
                                  className="glyphicon glyphicon-question-sign icon-black"
                                />
                              </Component>
                               Documentation
                            </a>
                          </SafeAnchor>
                        </li>
                      </MenuItem>
                    </Component>
                  </ul>
                </RootCloseWrapper>
              </DropdownMenu>
            </div>
          </ButtonGroup>
        </Dropdown>
      </Uncontrolled(Dropdown)>
    </BookmarkContainer>
  </Connect(BookmarkContainer)>
</Provider>
`;
